geode-core/src/distributedTest/java/org/apache/geode/distributed/internal/BadCacheLoaderDUnitTest.java

```diff
@@ -0,0 +1,100 @@
+
+import org.assertj.core.api.Assertions;
+import org.jetbrains.annotations.NotNull;
+import org.junit.Rule;
+import org.junit.Test;
+
+import org.apache.geode.InternalGemFireException;
+import org.apache.geode.cache.Cache;
+import org.apache.geode.cache.Region;
+import org.apache.geode.cache.RegionShortcut;
+import org.apache.geode.distributed.ConfigurationProperties;
+import org.apache.geode.test.dunit.SerializableRunnableIF;
+import org.apache.geode.test.dunit.VM;
+import org.apache.geode.test.dunit.rules.CacheRule;
+import org.apache.geode.test.dunit.rules.DistributedRule;
+
+public class BadCacheLoaderDUnitTest implements Serializable {
+  public static final String TEST_REGION = "testRegion";
+  public static final String TEST_KEY = "testKey";
+  public static final String TEST_VALUE = "testValue";
+
+  static class NotSerializableTestException extends RuntimeException {
+    Object unserializableField = new Object();
+  }
+
+  @Rule
+  public DistributedRule distributedRule = new DistributedRule(2);
+
+  CacheRule cacheRule = CacheRule.builder().build();
+
+  /**
+   * Ensure that a cache loader throwing an exception that is not serializable is handled
+   * correctly
+   */
+  @Test
+  public void testNonSerializableObjectReturnedByCacheLoader() throws Exception {
+    final VM cacheLoaderVM = VM.getVM(0);
+    final VM fetchingVM = VM.getVM(1);
+
+    final Properties properties = new Properties();
+    properties.put(ConfigurationProperties.SERIALIZABLE_OBJECT_FILTER,
+        NotSerializableTestException.class.getName());
+
+    cacheLoaderVM.invoke("create a region with a bad cache loader",
+        createRegionWithBadCacheLoader(properties));
+
+    Assertions.assertThatThrownBy(() -> fetchingVM.invoke("fetch something from the cache",
+        fetchValueCausingCacheLoad(properties)))
+        .hasCauseInstanceOf(InternalGemFireException.class)
+        .hasRootCauseInstanceOf(NotSerializableException.class)
+        .hasRootCauseMessage("java.lang.Object");
+  }
+
+  @NotNull
+  private SerializableRunnableIF fetchValueCausingCacheLoad(Properties properties) {
+    return () -> {
+      final Cache cache = cacheRule.getOrCreateCache(properties);
+      final Region<String, Object> testRegion =
+          cache.<String, Object>createRegionFactory(RegionShortcut.PARTITION)
+              .setCacheLoader(helper -> new Object())
+              .create(TEST_REGION);
+      testRegion.getAttributesMutator().setCacheLoader(helper -> "should not be invoked");
+      testRegion.get(TEST_KEY);
+    };
+  }
+
+  @NotNull
+  private SerializableRunnableIF createRegionWithBadCacheLoader(Properties properties) {
+    return () -> {
+      final Cache cache = cacheRule.getOrCreateCache(properties);
+      final Region<String, Object> testRegion =
+          cache.<String, Object>createRegionFactory(RegionShortcut.PARTITION)
+              .setCacheLoader(helper -> {
+                throw new NotSerializableTestException();
```

Comment:
it's so cool that it's possible to inject this lambda here that simply throws the exception